### DISCUSSION TITLE
Oauth2 driver exception in oauthCallback

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#11048] [Archive]
Hi,

Here is the error I receive when I authenticate with Facebook. [method throwing the exception](https://github.com/directus/directus/blob/7833160348967e0b5d61d29965bac3fb4d25a519/api/src/auth/drivers/oauth2.ts#L103) is `client.authCallback`
```
RPError: id_token detected in the response, you must use client.callback() instead of client.oauthCallback
```

I do not know yet from when exactly I have this issue.
I just tried to trace it and found the following:

This PR [Update dependency openid-client to v5](https://github.com/directus/directus/pull/9202) seems to introduce the issue.

[The release 5.1](https://github.com/panva/node-openid-client/compare/v5.0.2...v5.1.0) from open-client add new check that is [rejecting oauthCallback when id_token is detected](https://github.com/panva/node-openid-client/commit/92ffee5c63dc31fb578c731572cf0f83a7b53f1d)

@aidenfoxx @rijkvanzanten Unless it comes from something else, could you please tell me if I should open an issue about that.

**Comment by aidenfoxx** [comment]
Is there any reason you're using the oauth provider for Facebook instead of openid? That should fix the issue.

It is a little bit disappointing that openid-client "fixed" this apparent issue, because it makes things much less flexible for us.

EDIT:

Oh yeah, I forgot about this stupid issue https://github.com/directus/directus/discussions/9744. Facebook doesn't support the "code" flow with openid, which is super dumb. Guess we'll have to add a workaround for this openid-client change.

**Reply by gkielwasser** [reply]
I was using oauth instead of openid since this discussion :) https://github.com/directus/directus/discussions/9744

working with oauth driver was my only working solution so far.

**Reply by aidenfoxx** [reply]
Started a discussion on their side to have the check removed https://github.com/panva/node-openid-client/discussions/459

**Reply by aidenfoxx** [reply]
@gkielwasser Does the issue persist if you remove "openid" from the scope?

**Reply by gkielwasser** [reply]
The scope I'm using is the following
```
AUTH_FACEBOOK_SCOPE="email public_profile"
```

**Reply by aidenfoxx** [reply]
@gkielwasser Their developer did some testing on their end and believes this error shouldn't occur as long as "openid" is not in the scope. Are you sure your config is correct?

**Comment by gkielwasser** [comment]
@aidenfoxx yep, at least it is what I set for AUTH_FACEBOOK_SCOPE env

I have openid set for AUTH_GOOGLE_DRIVER but that shouldn't matter right

**Reply by gkielwasser** [reply]
I just tried reproducing the issue but I'm afraid I cannot anymore. I still have openid-client v5.1.1

@aidenfoxx honestly I don't get it, it do not have anymore the error thrown when I login. I didn't do further investigations since I have open this discussions..

I will still performs tests tonight and tomorrow to confirm.

**Reply by gkielwasser** [reply]
@aidenfoxx Ok my previous message is incorrect. It is working because I didn't revert the v4.9.1 of openid-client.
At least this confirms it works with this version,

**Reply by gkielwasser** [reply]
@aidenfoxx I still have the same error behavior with v5.1.1

**Reply by aidenfoxx** [reply]
I'll try and figure it out over the weekend.

**Reply by gkielwasser** [reply]
Thanks for your time! 

I hope this will help debugging. I have located more accuratly which line throw the error. There are 2 places where it could throw, [it is the second place](https://github.com/panva/node-openid-client/blob/5120a076d0b5b24b9ebd0dcdb8b40d4dfcd535a3/lib/client.js#L623).

```
if (params.code) {
      const tokenset = await this.grant(
        {
          ...exchangeBody,
          grant_type: 'authorization_code',
          code: params.code,
          redirect_uri: redirectUri,
          code_verifier: checks.code_verifier,
        },
        { clientAssertionPayload, DPoP },
      );

  ====>    if ('id_token' in tokenset) {
        throw new RPError({
          message:
            'id_token detected in the response, you must use client.callback() instead of client.oauthCallback()',
          params,
        });
      }
```

**Reply by aidenfoxx** [reply]
@gkielwasser So having debugged the issue both you and the module developer are right! The issue is that you've authenticated with Facebook with the "openid" scope in the past. Facebook retains this and always sends an "id_token" for every request after that. If you revoke access in your Facebook account and authenticate again without "openid" in the scope, everything should work as expected.

Frustrating, I know... but I don't think we'll be convincing the "openid-client" dev to reverse their commit.

**Reply by gkielwasser** [reply]
> @gkielwasser So having debugged the issue both you and the module developer are right! The issue is that you've authenticated with Facebook with the "openid" scope in the past. Facebook retains this and always sends an "id_token" for every request after that. If you revoke access in your Facebook account and authenticate again without "openid" in the scope, everything should work as expected.
> 
> Frustrating, I know... but I don't think we'll be convincing the "openid-client" dev to reverse their commit.

@aidenfoxx That was it ! Being in development I had no issue revoking the app access.
Thank you so much for help

**Reply by MalvinJay** [reply]
@aidenfoxx Please do you mind throwing more light on how to revoke the app access. In my situation, I am the web developer on a project that also has a mobile app developer and the both of us are suppose to implement Facebook login. At the moment, we use the same App created on Facebeook Developer Account and I'm thinking this OpenID connect is getting implemented from the mobile side of login. My question is, do I necessarily have to use this same Facebeook Developer Account App or I can create a new App and implement (block) OpenId access for my web logins to go work. Pardon me if I'm not making any sense but I'm in a tight corner now. By the way I use next-auth to implement the facebook login

**Comment by gkielwasser** [comment]
As suggested by @aidenfoxx revoking access was the solution
https://github.com/directus/directus/discussions/11048#discussioncomment-1977975

**Comment by sevetseh28** [comment]
I'm having the exact same issue when using AWS Cognito with Oauth flow to authenticate users. From raw logs:


directus    | {"level":10,"time":1661187374915,"pid":41,"hostname":"5ad812f377bf","params":{"code":"...","state":"..."},"name":"RPError","stack":"RPError: id_token detected in the response, you must use client.callback() instead of client.oauthCallback()\n    at Client.oauthCallback (/directus/node_modules/openid-client/lib/client.js:613:15)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async OAuth2AuthDriver.getUserID (/directus/node_modules/directus/dist/auth/drivers/oauth2.js:91:24)\n    at async AuthenticationService.login (/directus/node_modules/directus/dist/services/authentication.js:47:28)\n    at async /directus/node_modules/directus/dist/auth/drivers/oauth2.js:235:28","type":"Error","msg":"[OAuth2] Unknown RP error"}

**Comment by sevetseh28** [comment]
And If I don't request the `openid` scope I explicitly get the error:

Access token does not contain openid scope","name":"OPError","stack":"OPError: invalid_token (Access token does not contain openid scope)\

**Reply by aidenfoxx** [reply]
Use the openid driver. You can't use an openid provider with the the oauth driver. According to Google the URL for the cognito openid `.well-known` is "https://cognito-idp.{region}.amazonaws.com/{userPoolId}/.well-known/openid-configuration".
